tests-e2e/fixture/fixture.go

```diff
@@ -0,0 +1,196 @@
+		if err != nil {
+			if apierr.IsNotFound(err) {
+				return true, nil
+			} else {
+				return false, err
+			}
+		}
+
+		return false, nil
+	}); err != nil {
+		return err
+	}
+
+	// Create the namespace again
+	_, err = kubeClientSet.CoreV1().Namespaces().Create(context.Background(), &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{
+		Name: GitOpsServiceE2ENamespace,
+		Labels: map[string]string{
+			"argocd.argoproj.io/managed-by": "gitops-service-argocd",
+		},
+	}}, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Wait for Argo CD to process the namespace, before we exit:
+	// - This helps us avoid a race condition where the namespace is created, but Argo CD has not yet
+	//   set up proper roles for it.
+
+	if err := wait.Poll(time.Second*1, time.Minute*2, func() (done bool, err error) {
+		var roleBindings *v1.RoleBindingList
+		roleBindings, err = kubeClientSet.RbacV1().RoleBindings(GitOpsServiceE2ENamespace).List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		count := 0
+		// Exit the poll loop when there exists at least one rolebinding containing 'gitops-service-argocd' in it's name
+		// This helps us avoid a race condition where the namespace is created, but Argo CD has not yet
+		// set up proper roles for it.
+		for _, item := range roleBindings.Items {
+			if strings.Contains(item.Name, "gitops-service-argocd-") {
+				count++
+			}
+		}
+
+		if count >= 2 {
+			// We expect at least 2 rolebindings
+			return true, nil
+		}
+
+		return false, nil
+	}); err != nil {
+		return err
+	}
+
+	if err := wait.Poll(time.Second*1, time.Minute*2, func() (done bool, err error) {
+		var roles *v1.RoleList
+		roles, err = kubeClientSet.RbacV1().Roles(GitOpsServiceE2ENamespace).List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		count := 0
+		// Exit the poll loop when there exists at least one rolebinding containing 'gitops-service-argocd' in it's name
+		// This helps us avoid a race condition where the namespace is created, but Argo CD has not yet
+		// set up proper roles for it.
+		for _, item := range roles.Items {
+			if strings.Contains(item.Name, "gitops-service-argocd-") {
+				count++
+			}
+		}
+
+		if count >= 2 {
+			// We expect at least 2 roles
+			return true, nil
+		}
+
+		return false, nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func GetKubeConfig() (*rest.Config, error) {
+
+	overrides := clientcmd.ConfigOverrides{}
+
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, &overrides, os.Stdin)
+
+	restConfig, err := clientConfig.ClientConfig()
+
+	if restConfig != nil {
+		// Sanity check that we're not running on a known staging system
+		if strings.Contains(restConfig.Host, "x99m.p1.openshiftapps.com") {
+			return nil, fmt.Errorf("E2E tests should not be run on staging server")
+		}
+	}
+
```

Comment:
> define an errGetKubeConfig and return this in case of an error, wrapping inside of it the errors coming from the system itself.

Not sure I see what the request is here, can you provide an example of what you mean?

> In case there is no error, print a statement it saying that kubeconfig is configured properly against X target cluster.

Tried this but too verbose, see above for result.

Reply:
> Not sure I see what the request is here, can you provide an example of what you mean?

```go
// errors per operation
var errClient = errors.New("couldn't get k8s client")
var errCreate = errors.New("couldn't create k8s resource")
var errDelete = errors.New("couldn't delete k8s object from the namespace")
var errYouMentionIt = errors.New("think of another error")

// Create creates the given K8s resource, returning an error on failure, or nil otherwise.
func Create(obj client.Object) error {
	// ~~~ setup ~~~ //
	kubeClient, err := fixture.GetKubeClient()
	if err != nil {
		return fmt.Errorf("%v (error: %w)", errClient, err)
	}

	// ~~~ test ~~~ //
	fmt.Println("Creating ", obj.GetName())

	if err := kubeClient.Create(context.Background(), obj); err != nil {
		return fmt.Errorf("%v (error: %w)", errCreate, err)
	}

	return nil
}
```

> 
> > In case there is no error, print a statement it saying that kubeconfig is configured properly against X target cluster.
> 
> Tried this but too verbose, see above for result.

Yes it's too verbose but for a food reason. As I said yesterday in the parking lot, in the end it's all about finding the right balance. If we have just a few tests written by us and maintained by us -- we are probably good to go without that level of verbosity.  But if we are going to be responsible to debug a few hundreds of tests written by a whole organization, then verbosity would be greatly appreciated. 

Another idea might be to consider introducing debug levels where you can have all the verbosity you need + run a few interesting kubectl commands (in case of failed testsuite) on the cluster itself (`oc get nodes`, `oc top`, etc), and finally export the output into multiple files and generate a compressed logs.tar.gz file for us. 

But anyway, we can discuss all those things later, not in this PR.

Reply:
@drpaneas: ah, that is a lot clearer for me, thanks!

Reply:
@drpaneas: And +1 for different debug log levels for our components (for example, the ability to enable debug for a particular file/component only), +1 for provided a more consistent log format (our current log format is very anarchic :smile: ).